File: game/src/World.cpp
#include "zomboid/World.hpp"

#include <algorithm>

namespace zb {

const std::vector<Avenue>& avenues() {
    // East/West streets (rows), north -> south.
    static const std::vector<Avenue> kAvenues = {
        {"Ship Creek", 6},
        {"1st Avenue", 14},
        {"3rd Avenue", 22},
        {"4th Avenue", 30},
        {"5th Avenue", 38},
        {"6th Avenue", 46},
        {"9th Avenue", 54},
        {"Delaney Park Strip", 62},
        {"15th Avenue", 72},
        {"Northern Lights Boulevard", 88},
        {"Benson Boulevard", 100},
        {"Tudor Road", 122},
    };
    return kAvenues;
}

const std::vector<Street>& streets() {
    // North/South streets (columns), west -> east.
    static const std::vector<Street> kStreets = {
        {"Cook Inlet", 4},
        {"L Street", 14},
        {"K Street", 24},
        {"I Street", 34},
        {"G Street", 44},
        {"E Street", 54},
        {"C Street", 66},
        {"A Street", 78},
        {"Cordova Street", 90},
        {"Gambell Street", 104},
        {"Ingra Street", 114},
        {"Lake Otis Parkway", 132},
        {"Boniface Parkway", 150},
    };
    return kStreets;
}

const std::vector<Building>& buildings() {
    // name, x, y, w, h, doorX, doorY, loot, kind
    static const std::vector<Building> kBuildings = {
        {"Hotel Captain Cook", 26, 24, 10, 5, 4, 4, 2, "hotel"},
        {"4th Avenue Theatre", 46, 24, 7, 4, 3, 3, 1, "theatre"},
        {"Anchorage City Hall", 56, 24, 6, 4, 2, 3, 1, "gov"},
        {"Egan Convention Center", 68, 24, 9, 4, 4, 3, 1, "civic"},
        {"Dena'ina Center", 26, 32, 8, 4, 3, 3, 1, "civic"},
        {"Town Square Park", 46, 32, 7, 5, 3, 0, 0, "park"},
        {"Performing Arts Center", 56, 32, 8, 5, 3, 4, 1, "civic"},
        {"Nordstrom (5th Ave Mall)", 68, 32, 9, 5, 4, 4, 3, "mall"},
        {"Anchorage Museum", 56, 40, 11, 5, 5, 0, 1, "museum"},
        {"Z.J. Loussac Library", 80, 40, 8, 5, 3, 4, 1, "gov"},
        {"Snow City Cafe", 16, 40, 5, 3, 2, 2, 2, "food"},
        {"Glacier Brewhouse", 26, 40, 5, 3, 2, 2, 2, "food"},
        {"49th State Brewing", 36, 40, 6, 3, 2, 2, 2, "food"},
        {"Carrs Aurora Village", 92, 32, 10, 6, 5, 5, 3, "grocery"},
        {"Fred Meyer", 106, 32, 12, 7, 6, 6, 3, "grocery"},
        {"Title Wave Books", 92, 42, 7, 4, 3, 3, 1, "shop"},
        {"REI Anchorage", 102, 42, 7, 4, 3, 3, 3, "outdoor"},
        {"Sullivan Arena", 100, 50, 12, 7, 6, 6, 1, "arena"},
        {"Merrill Field Hangars", 120, 24, 14, 8, 7, 7, 2, "industrial"},
        {"Providence Medical Center", 120, 92, 16, 9, 8, 8, 3, "hospital"},
        {"Alaska Native Medical Center", 120, 104, 14, 8, 7, 7, 3, "hospital"},
        {"Moose's Tooth Pub", 96, 92, 6, 4, 3, 3, 2, "food"},
        {"Bear Tooth Theatrepub", 84, 92, 6, 4, 3, 3, 2, "food"},
        {"Chilkoot Charlie's", 16, 90, 7, 4, 3, 3, 2, "bar"},
        {"Spenard Builders Supply", 16, 100, 9, 5, 4, 4, 3, "hardware"},
        {"Dimond Center Mall", 30, 118, 16, 9, 8, 8, 3, "mall"},
        {"University of Alaska Anchorage", 132, 70, 16, 10, 8, 9, 2, "campus"},
        {"Alaska Regional Hospital", 132, 56, 12, 7, 6, 6, 3, "hospital"},
        {"Anchorage Police Dept", 80, 56, 8, 5, 4, 4, 2, "police"},
        {"APD Crime Lab", 92, 56, 6, 4, 2, 3, 2, "police"},
        {"Alaska Railroad Depot", 26, 8, 10, 4, 5, 3, 1, "transit"},
        {"Westchester Lagoon Pavilion", 14, 64, 5, 3, 2, 2, 0, "park"},
        {"Bass Pro / Sportsman", 100, 100, 10, 6, 5, 5, 3, "outdoor"},
        {"Midtown Mall", 70, 92, 9, 5, 4, 4, 3, "mall"},
        {"The Lakefront Hotel", 54, 100, 9, 5, 4, 4, 2, "hotel"},
        {"Tesoro Gas Station", 60, 76, 5, 3, 2, 2, 2, "gas"},
        {"Holiday Fuel", 96, 76, 5, 3, 2, 2, 2, "gas"},
    };
    return kBuildings;
}

namespace {

bool inBounds(int x, int y) { return x >= 0 && y >= 0 && x < kMapW && y < kMapH; }

} // namespace

World generateWorld() {
    World world;
    world.w = kMapW;
    world.h = kMapH;
    world.tiles.assign(static_cast<size_t>(kMapW) * static_cast<size_t>(kMapH), Tile::Grass);

    // --- Cook Inlet water on the far west; ragged shoreline ---
    for (int y = 0; y < kMapH; y++) {
        for (int x = 0; x < 4; x++) world.set(x, y, Tile::Water);
        if ((y * 7) % 5 == 0) world.set(4, y, Tile::Water);
    }
    // Ship Creek strip across the north.
    for (int x = 4; x < kMapW; x++) {
        world.set(x, 2, Tile::Water);
        world.set(x, 3, Tile::Water);
    }
    // Chugach treeline / greenbelt on the east edge.
    for (int y = 0; y < kMapH; y++) {
        for (int x = kMapW - 6; x < kMapW; x++) {
            if ((x * 3 + y * 5) % 4 != 0)
                world.set(x, y, Tile::Tree);
            else
                world.set(x, y, Tile::Grass);
        }
    }

    // --- Avenues (horizontal streets) ---
    for (const auto& av : avenues()) {
        const int r = av.row;
        for (int x = 5; x < kMapW - 6; x++) {
            if (inBounds(x, r - 1)) world.set(x, r - 1, Tile::Sidewalk);
            if (inBounds(x, r)) world.set(x, r, Tile::Street);
            if (inBounds(x, r + 1)) world.set(x, r + 1, Tile::Street);
            if (inBounds(x, r + 2)) world.set(x, r + 2, Tile::Sidewalk);
        }
    }

    // Ship Creek railyard.
    for (int x = 5; x < kMapW - 6; x++) {
        world.set(x, 8, Tile::Rail);
        world.set(x, 9, Tile::Rail);
    }

    // --- Streets (vertical) ---
    for (const auto& st : streets()) {
        const int c = st.col;
        for (int y = 5; y < kMapH - 2; y++) {
            if (inBounds(c - 1, y)) world.set(c - 1, y, Tile::Sidewalk);
            if (inBounds(c, y)) world.set(c, y, Tile::Street);
            if (inBounds(c + 1, y)) world.set(c + 1, y, Tile::Street);
            if (inBounds(c + 2, y)) world.set(c + 2, y, Tile::Sidewalk);
        }
    }

    // --- Buildings ---
    for (const auto& b : buildings()) {
        const bool isPark = b.kind == "park";
        for (int yy = 0; yy < b.h; yy++) {
            for (int xx = 0; xx < b.w; xx++) {
                const int gx = b.x + xx, gy = b.y + yy;
                if (!inBounds(gx, gy)) continue;
                const bool edge = xx == 0 || yy == 0 || xx == b.w - 1 || yy == b.h - 1;
                if (isPark) {
                    world.set(gx, gy, (xx + yy) % 3 == 0 ? Tile::Tree : Tile::Grass);
                } else if (edge) {
                    world.set(gx, gy, Tile::Wall);
                } else {
                    world.set(gx, gy, Tile::Floor);
                }
            }
        }
        // Door + apron/parking lot in front.
        if (!isPark) {
            const int dx = b.x + b.doorX, dy = b.y + b.doorY;
            if (inBounds(dx, dy)) world.set(dx, dy, Tile::Door);
            for (int k = 1; k <= 2; k++) {
                const int ay = dy + k;
                if (inBounds(dx, ay) && world.at(dx, ay) == Tile::Grass) world.set(dx, ay, Tile::Lot);
            }
        }
        // Loot containers scattered inside; count scales with richness.
        const int nContainers = b.loot + 1;
        for (int i = 0; i < nContainers; i++) {
            const int cx = b.x + 1 + ((i * 3 + 1) % std::max(1, b.w - 2));
            const int cy = b.y + 1 + ((i * 2 + 1) % std::max(1, b.h - 2));
            if (inBounds(cx, cy) && world.at(cx, cy) == Tile::Floor) {
                world.containers.push_back(Container{cx, cy, b.loot, b.name, b.kind, false});
            }
        }
    }

    // --- Zombie spawn points: doorways + intersections ---
    for (const auto& b : buildings()) {
        world.spawnPoints.push_back(SpawnPoint{b.x + b.doorX, b.y + b.doorY + 1});
    }
    for (const auto& av : avenues()) {
        for (const auto& st : streets()) {
            world.spawnPoints.push_back(SpawnPoint{st.col, av.row});
        }
    }

    world.spawn = Vec2{49.0f, 36.0f};
    return world;
}

} // namespace zb

File: zomboid/js/world.js
/* =============================================================================
 *  ZOMBOID: ANCHORAGE  —  WORLD DATA
 *  A tile map of downtown / midtown Anchorage, Alaska.
 *  Street grid + named real-world buildings & landmarks.
 *  Tile codes:
 *    0 grass        1 street(asphalt)   2 sidewalk     3 building floor
 *    4 wall         5 door              6 water        7 tree
 *    8 parkingLot   9 railroad
 * ========================================================================== */
(function (global) {
  'use strict';

  const T = {
    GRASS: 0, STREET: 1, SIDEWALK: 2, FLOOR: 3, WALL: 4,
    DOOR: 5, WATER: 6, TREE: 7, LOT: 8, RAIL: 9,
  };

  const MAP_W = 160;
  const MAP_H = 140;

  // ---- East/West streets (rows) of Anchorage, north -> south ----
  // Ship Creek / railyard sits at the very north, then the numbered avenues.
  const AVENUES = [
    { name: 'Ship Creek',                 row: 6  },
    { name: '1st Avenue',                 row: 14 },
    { name: '3rd Avenue',                 row: 22 },
    { name: '4th Avenue',                 row: 30 },
    { name: '5th Avenue',                 row: 38 },
    { name: '6th Avenue',                 row: 46 },
    { name: '9th Avenue',                 row: 54 },
    { name: 'Delaney Park Strip',         row: 62 },
    { name: '15th Avenue',                row: 72 },
    { name: 'Northern Lights Boulevard',  row: 88 },
    { name: 'Benson Boulevard',           row: 100 },
    { name: 'Tudor Road',                 row: 122 },
  ];

  // ---- North/South streets (columns), west -> east ----
  const STREETS = [
    { name: 'Cook Inlet',          col: 4   },
    { name: 'L Street',            col: 14  },
    { name: 'K Street',            col: 24  },
    { name: 'I Street',            col: 34  },
    { name: 'G Street',            col: 44  },
    { name: 'E Street',            col: 54  },
    { name: 'C Street',            col: 66  },
    { name: 'A Street',            col: 78  },
    { name: 'Cordova Street',      col: 90  },
    { name: 'Gambell Street',      col: 104 },
    { name: 'Ingra Street',        col: 114 },
    { name: 'Lake Otis Parkway',   col: 132 },
    { name: 'Boniface Parkway',    col: 150 },
  ];

  // ---- Named landmark buildings. x,y = top-left tile; w,h in tiles ----
  // door = [dx,dy] offset for the entrance. loot = container richness 0..3
  const BUILDINGS = [
    { name: 'Hotel Captain Cook',            x: 26, y: 24, w: 10, h: 5, door: [4, 4], loot: 2, kind: 'hotel' },
    { name: '4th Avenue Theatre',            x: 46, y: 24, w: 7,  h: 4, door: [3, 3], loot: 1, kind: 'theatre' },
    { name: 'Anchorage City Hall',           x: 56, y: 24, w: 6,  h: 4, door: [2, 3], loot: 1, kind: 'gov' },
    { name: 'Egan Convention Center',        x: 68, y: 24, w: 9,  h: 4, door: [4, 3], loot: 1, kind: 'civic' },
    { name: 'Dena’ina Center',          x: 26, y: 32, w: 8,  h: 4, door: [3, 3], loot: 1, kind: 'civic' },
    { name: 'Town Square Park',              x: 46, y: 32, w: 7,  h: 5, door: [3, 0], loot: 0, kind: 'park' },
    { name: 'Performing Arts Center',        x: 56, y: 32, w: 8,  h: 5, door: [3, 4], loot: 1, kind: 'civic' },
    { name: 'Nordstrom (5th Ave Mall)',      x: 68, y: 32, w: 9,  h: 5, door: [4, 4], loot: 3, kind: 'mall' },
    { name: 'Anchorage Museum',              x: 56, y: 40, w: 11, h: 5, door: [5, 0], loot: 1, kind: 'museum' },
    { name: 'Z.J. Loussac Library',          x: 80, y: 40, w: 8,  h: 5, door: [3, 4], loot: 1, kind: 'gov' },
    { name: 'Snow City Cafe',                x: 16, y: 40, w: 5,  h: 3, door: [2, 2], loot: 2, kind: 'food' },
    { name: 'Glacier Brewhouse',             x: 26, y: 40, w: 5,  h: 3, door: [2, 2], loot: 2, kind: 'food' },
    { name: '49th State Brewing',            x: 36, y: 40, w: 6,  h: 3, door: [2, 2], loot: 2, kind: 'food' },
    { name: 'Carrs Aurora Village',          x: 92, y: 32, w: 10, h: 6, door: [5, 5], loot: 3, kind: 'grocery' },
    { name: 'Fred Meyer',                    x: 106,y: 32, w: 12, h: 7, door: [6, 6], loot: 3, kind: 'grocery' },
    { name: 'Title Wave Books',              x: 92, y: 42, w: 7,  h: 4, door: [3, 3], loot: 1, kind: 'shop' },
    { name: 'REI Anchorage',                 x: 102,y: 42, w: 7,  h: 4, door: [3, 3], loot: 3, kind: 'outdoor' },
    { name: 'Sullivan Arena',                x: 100,y: 50, w: 12, h: 7, door: [6, 6], loot: 1, kind: 'arena' },
    { name: 'Merrill Field Hangars',         x: 120,y: 24, w: 14, h: 8, door: [7, 7], loot: 2, kind: 'industrial' },
    { name: 'Providence Medical Center',     x: 120,y: 92, w: 16, h: 9, door: [8, 8], loot: 3, kind: 'hospital' },
    { name: 'Alaska Native Medical Center',  x: 120,y: 104,w: 14, h: 8, door: [7, 7], loot: 3, kind: 'hospital' },
    { name: 'Moose’s Tooth Pub',        x: 96, y: 92, w: 6,  h: 4, door: [3, 3], loot: 2, kind: 'food' },
    { name: 'Bear Tooth Theatrepub',         x: 84, y: 92, w: 6,  h: 4, door: [3, 3], loot: 2, kind: 'food' },
    { name: 'Chilkoot Charlie’s',       x: 16, y: 90, w: 7,  h: 4, door: [3, 3], loot: 2, kind: 'bar' },
    { name: 'Spenard Builders Supply',       x: 16, y: 100,w: 9,  h: 5, door: [4, 4], loot: 3, kind: 'hardware' },
    { name: 'Dimond Center Mall',            x: 30, y: 118,w: 16, h: 9, door: [8, 8], loot: 3, kind: 'mall' },
    { name: 'University of Alaska Anchorage', x: 132,y: 70, w: 16, h: 10,door: [8, 9], loot: 2, kind: 'campus' },
    { name: 'Alaska Regional Hospital',      x: 132,y: 56, w: 12, h: 7, door: [6, 6], loot: 3, kind: 'hospital' },
    { name: 'Anchorage Police Dept',         x: 80, y: 56, w: 8,  h: 5, door: [4, 4], loot: 2, kind: 'police' },
    { name: 'APD Crime Lab',                 x: 92, y: 56, w: 6,  h: 4, door: [2, 3], loot: 2, kind: 'police' },
    { name: 'Alaska Railroad Depot',         x: 26, y: 8,  w: 10, h: 4, door: [5, 3], loot: 1, kind: 'transit' },
    { name: 'Westchester Lagoon Pavilion',   x: 14, y: 64, w: 5,  h: 3, door: [2, 2], loot: 0, kind: 'park' },
    { name: 'Bass Pro / Sportsman',          x: 100,y: 100,w: 10, h: 6, door: [5, 5], loot: 3, kind: 'outdoor' },
    { name: 'Midtown Mall',                  x: 70, y: 92, w: 9,  h: 5, door: [4, 4], loot: 3, kind: 'mall' },
    { name: 'The Lakefront Hotel',           x: 54, y: 100,w: 9,  h: 5, door: [4, 4], loot: 2, kind: 'hotel' },
    { name: 'Tesoro Gas Station',            x: 60, y: 76, w: 5,  h: 3, door: [2, 2], loot: 2, kind: 'gas' },
    { name: 'Holiday Fuel',                  x: 96, y: 76, w: 5,  h: 3, door: [2, 2], loot: 2, kind: 'gas' },
  ];

  global.ANCHORAGE = { T, MAP_W, MAP_H, AVENUES, STREETS, BUILDINGS };
})(typeof window !== 'undefined' ? window : this);
